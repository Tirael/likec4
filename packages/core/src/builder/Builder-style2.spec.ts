import { describe, expect, it } from 'vitest'
import { Builder } from './Builder'

describe('Builder (style 2)', () => {
  const spec = Builder
    .specification({
      elements: {
        system: {
          style: {
            size: 'lg',
            textSize: 'sm',
          },
        },
        component: {},
        actor: {},
      },
      deployments: {
        env: {
          style: {
            size: 'lg',
          },
        },
        node: {},
      },
    })

  it('should build ', () => {
    const b = spec.clone()
      .model(({ system, actor, component }, _) =>
        _(
          actor('customer'),
          system('cloud').with(
            component('ui'),
          ),
        )
      )
      .deployment(({ env, node, instanceOf }, _) =>
        _(
          env('prod').with(
            node('eu').with(
              instanceOf('cloud.ui'),
            ),
          ),
          env('dev'),
          node('dev.local'),
        )
      )
      .views(({ view, viewOf, deploymentView, $include }, _) =>
        _(
          view('index', 'Index').with(
            $include('cloud.*'),
          ),
          viewOf('cloud', 'cloud.ui').with(
            $include('* -> cloud.**'),
          ),
          deploymentView('deployment', 'Deployment').with(
            $include('prod.**'),
          ),
        )
      )

    expect(b.build()).toMatchSnapshot()
  })

  it('should fail if invalid ID provided ', () => {
    expect(() => {
      spec.model(({ actor }, _) =>
        _(
          actor('cust.omer'),
        )
      )
    }).toThrowError('Parent element with id "cust" not found for element with id "cust.omer"')
  })

  it('should fail on invalid instance ', () => {
    const b = spec.clone()
      .model(_ =>
        _.model(
          _.component('cloud'),
          _.component('cloud.ui'),
        )
      )

    expect(() => {
      const raw = b.deployment(_ =>
        _.deployment(
          _.instanceOf('cloud.ui'),
        )
      ).build()
    }).toThrowError('Instance ui of cloud.ui must be deployed under a parent node')

    // Nested instanceOf is correct

    const raw = b.deployment(_ =>
      _.deployment(
        _.node('node').with(
          _.instanceOf('cloud.ui'),
        ),
      )
    ).build()
    expect(raw.deployments.elements).toEqual({
      node: expect.objectContaining({ id: 'node' }),
      // Take name from element
      'node.ui': expect.objectContaining({ id: 'node.ui', element: 'cloud.ui' }),
    })
  })

  it('should build activities', ({ expect }) => {
    const b = spec.clone()
      .model(_ =>
        _.model(
          _.component('s1').with(
            _.activity('A'),
            _.component('c1'),
          ),
          _.component('s2').with(
            _.activity('B'),
            _.component('c2'),
          ),
          _.activity('s2.c2#C'),
        )
      )

    expect(b.build()).toMatchSnapshot()
  })

  it('should build activities with title', ({ expect }) => {
    expect(
      spec.clone()
        .model(_ =>
          _.model(
            _.component('s1').with(
              _.activity('A', {
                title: 'Title A',
              }),
            ),
          )
        ).build(),
    ).toMatchSnapshot()
  })

  it('should build activities with steps (array)', ({ expect }) => {
    expect(
      spec.clone()
        .model(_ =>
          _.model(
            _.component('s1').with(
              _.component('c1'),
            ),
            _.component('s2').with(
              _.component('c2'),
            ),
            _.activity('s2.c2#C', [
              _.step('-> s1.c1'),
              _.step('<- s1.c1'),
            ]),
          )
        ).build(),
    ).toMatchSnapshot()
  })

  it('should build activities with steps (string array)', ({ expect }) => {
    expect(
      spec.clone().model(_ =>
        _.model(
          _.component('s1'),
          _.component('s2').with(
            _.activity('B'),
          ),
          _.activity('s1#A', [
            '-> s2#B',
          ]),
        )
      ).build(),
    ).toMatchSnapshot()
  })

  it('should build activities with steps (object)', ({ expect }) => {
    expect(
      spec.clone().model(_ =>
        _.model(
          _.component('s1'),
          _.component('s2').with(
            _.activity('B'),
          ),
          _.activity('s1#A', {
            steps: [
              ['-> s2#B', 'title1'],
              ['<- s2#B', {
                title: 'title2',
              }],
            ],
          }),
        )
      ).build(),
    ).toMatchSnapshot()
  })
})
