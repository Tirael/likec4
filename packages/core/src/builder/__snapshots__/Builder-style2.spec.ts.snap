// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Builder (style 2) > should build  1`] = `
{
  "activities": {},
  "deployments": {
    "elements": {
      "dev": {
        "color": "primary",
        "description": null,
        "id": "dev",
        "kind": "env",
        "shape": "rectangle",
        "style": {
          "size": "lg",
        },
        "tags": null,
        "technology": null,
        "title": "dev",
      },
      "dev.local": {
        "color": "primary",
        "description": null,
        "id": "dev.local",
        "kind": "node",
        "shape": "rectangle",
        "style": {},
        "tags": null,
        "technology": null,
        "title": "local",
      },
      "prod": {
        "color": "primary",
        "description": null,
        "id": "prod",
        "kind": "env",
        "shape": "rectangle",
        "style": {
          "size": "lg",
        },
        "tags": null,
        "technology": null,
        "title": "prod",
      },
      "prod.eu": {
        "color": "primary",
        "description": null,
        "id": "prod.eu",
        "kind": "node",
        "shape": "rectangle",
        "style": {},
        "tags": null,
        "technology": null,
        "title": "eu",
      },
      "prod.eu.ui": {
        "element": "cloud.ui",
        "id": "prod.eu.ui",
      },
    },
    "relations": {},
  },
  "elements": {
    "cloud": {
      "color": "primary",
      "description": null,
      "id": "cloud",
      "kind": "system",
      "links": null,
      "shape": "rectangle",
      "style": {
        "size": "lg",
        "textSize": "sm",
      },
      "tags": null,
      "technology": null,
      "title": "cloud",
    },
    "cloud.ui": {
      "color": "primary",
      "description": null,
      "id": "cloud.ui",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "ui",
    },
    "customer": {
      "color": "primary",
      "description": null,
      "id": "customer",
      "kind": "actor",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "customer",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {
      "env": {
        "style": {
          "size": "lg",
        },
      },
      "node": {},
    },
    "elements": {
      "actor": {},
      "component": {},
      "system": {
        "style": {
          "size": "lg",
          "textSize": "sm",
        },
      },
    },
    "relationships": {},
    "tags": [],
  },
  "views": {
    "cloud": {
      "__": "element",
      "customColorDefinitions": {},
      "description": null,
      "id": "cloud",
      "links": null,
      "rules": [
        {
          "include": [
            {
              "source": {
                "wildcard": true,
              },
              "target": {
                "ref": {
                  "model": "cloud",
                },
                "selector": "descendants",
              },
            },
          ],
        },
      ],
      "tags": null,
      "title": null,
      "viewOf": "cloud.ui",
    },
    "deployment": {
      "__": "deployment",
      "customColorDefinitions": {},
      "description": null,
      "id": "deployment",
      "links": null,
      "rules": [
        {
          "include": [
            {
              "ref": {
                "deployment": "prod",
              },
              "selector": "descendants",
            },
          ],
        },
      ],
      "tags": null,
      "title": "Deployment",
    },
    "index": {
      "__": "element",
      "customColorDefinitions": {},
      "description": null,
      "id": "index",
      "links": null,
      "rules": [
        {
          "include": [
            {
              "ref": {
                "model": "cloud",
              },
              "selector": "children",
            },
          ],
        },
      ],
      "tags": null,
      "title": "Index",
    },
  },
}
`;

exports[`Builder (style 2) > should build activities 1`] = `
{
  "activities": {
    "s1#A": {
      "id": "s1#A",
      "modelRef": "s1",
      "name": "A",
      "steps": [],
    },
    "s2#B": {
      "id": "s2#B",
      "modelRef": "s2",
      "name": "B",
      "steps": [],
    },
    "s2.c2#C": {
      "id": "s2.c2#C",
      "modelRef": "s2.c2",
      "name": "C",
      "steps": [],
    },
  },
  "deployments": {
    "elements": {},
    "relations": {},
  },
  "elements": {
    "s1": {
      "color": "primary",
      "description": null,
      "id": "s1",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s1",
    },
    "s1.c1": {
      "color": "primary",
      "description": null,
      "id": "s1.c1",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "c1",
    },
    "s2": {
      "color": "primary",
      "description": null,
      "id": "s2",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s2",
    },
    "s2.c2": {
      "color": "primary",
      "description": null,
      "id": "s2.c2",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "c2",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {
      "env": {
        "style": {
          "size": "lg",
        },
      },
      "node": {},
    },
    "elements": {
      "actor": {},
      "component": {},
      "system": {
        "style": {
          "size": "lg",
          "textSize": "sm",
        },
      },
    },
    "relationships": {},
    "tags": [],
  },
  "views": {},
}
`;

exports[`Builder (style 2) > should build activities with steps (array) 1`] = `
{
  "activities": {
    "s2.c2#C": {
      "id": "s2.c2#C",
      "modelRef": "s2.c2",
      "name": "C",
      "steps": [
        {
          "id": "s2.c2#C_step1",
          "isBackward": false,
          "target": {
            "model": "s1.c1",
          },
        },
        {
          "id": "s2.c2#C_step2",
          "isBackward": true,
          "target": {
            "model": "s1.c1",
          },
        },
      ],
    },
  },
  "deployments": {
    "elements": {},
    "relations": {},
  },
  "elements": {
    "s1": {
      "color": "primary",
      "description": null,
      "id": "s1",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s1",
    },
    "s1.c1": {
      "color": "primary",
      "description": null,
      "id": "s1.c1",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "c1",
    },
    "s2": {
      "color": "primary",
      "description": null,
      "id": "s2",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s2",
    },
    "s2.c2": {
      "color": "primary",
      "description": null,
      "id": "s2.c2",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "c2",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {
      "env": {
        "style": {
          "size": "lg",
        },
      },
      "node": {},
    },
    "elements": {
      "actor": {},
      "component": {},
      "system": {
        "style": {
          "size": "lg",
          "textSize": "sm",
        },
      },
    },
    "relationships": {},
    "tags": [],
  },
  "views": {},
}
`;

exports[`Builder (style 2) > should build activities with steps (object) 1`] = `
{
  "activities": {
    "s1#A": {
      "id": "s1#A",
      "modelRef": "s1",
      "name": "A",
      "steps": [
        {
          "id": "s1#A_step1",
          "isBackward": false,
          "target": {
            "activity": "s2#B",
          },
          "title": "title1",
        },
        {
          "id": "s1#A_step2",
          "isBackward": true,
          "target": {
            "activity": "s2#B",
          },
          "title": "title2",
        },
      ],
    },
    "s2#B": {
      "id": "s2#B",
      "modelRef": "s2",
      "name": "B",
      "steps": [],
    },
  },
  "deployments": {
    "elements": {},
    "relations": {},
  },
  "elements": {
    "s1": {
      "color": "primary",
      "description": null,
      "id": "s1",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s1",
    },
    "s2": {
      "color": "primary",
      "description": null,
      "id": "s2",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s2",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {
      "env": {
        "style": {
          "size": "lg",
        },
      },
      "node": {},
    },
    "elements": {
      "actor": {},
      "component": {},
      "system": {
        "style": {
          "size": "lg",
          "textSize": "sm",
        },
      },
    },
    "relationships": {},
    "tags": [],
  },
  "views": {},
}
`;

exports[`Builder (style 2) > should build activities with steps (string array) 1`] = `
{
  "activities": {
    "s1#A": {
      "id": "s1#A",
      "modelRef": "s1",
      "name": "A",
      "steps": [
        {
          "id": "s1#A_step1",
          "isBackward": false,
          "target": {
            "activity": "s2#B",
          },
        },
      ],
    },
    "s2#B": {
      "id": "s2#B",
      "modelRef": "s2",
      "name": "B",
      "steps": [],
    },
  },
  "deployments": {
    "elements": {},
    "relations": {},
  },
  "elements": {
    "s1": {
      "color": "primary",
      "description": null,
      "id": "s1",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s1",
    },
    "s2": {
      "color": "primary",
      "description": null,
      "id": "s2",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s2",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {
      "env": {
        "style": {
          "size": "lg",
        },
      },
      "node": {},
    },
    "elements": {
      "actor": {},
      "component": {},
      "system": {
        "style": {
          "size": "lg",
          "textSize": "sm",
        },
      },
    },
    "relationships": {},
    "tags": [],
  },
  "views": {},
}
`;

exports[`Builder (style 2) > should build activities with title 1`] = `
{
  "activities": {
    "s1#A": {
      "id": "s1#A",
      "modelRef": "s1",
      "name": "A",
      "steps": [],
      "title": "Title A",
    },
  },
  "deployments": {
    "elements": {},
    "relations": {},
  },
  "elements": {
    "s1": {
      "color": "primary",
      "description": null,
      "id": "s1",
      "kind": "component",
      "links": null,
      "shape": "rectangle",
      "style": {},
      "tags": null,
      "technology": null,
      "title": "s1",
    },
  },
  "globals": {
    "dynamicPredicates": {},
    "predicates": {},
    "styles": {},
  },
  "imports": {},
  "relations": {},
  "specification": {
    "deployments": {
      "env": {
        "style": {
          "size": "lg",
        },
      },
      "node": {},
    },
    "elements": {
      "actor": {},
      "component": {},
      "system": {
        "style": {
          "size": "lg",
          "textSize": "sm",
        },
      },
    },
    "relationships": {},
    "tags": [],
  },
  "views": {},
}
`;
